### ISSUE TITLE
Incorrect lifetime used for mutable references to trait objects (correct one is used for non-mut references)

### ISSUE THREAD

**Issue by VFLashM** [rust-lang/rust#72337]
Instead of explicitly defined lifetime, rust uses anonymous lifetime of surrounding block.
For some reason problem only happens for mut references, non-mut references work just fine.

I tried this code:

```rust
// trait allowing to get reference to stored i32
trait T {
    fn get(&self) -> &i32;
    fn get_mut(&mut self) -> &mut i32;
}

// iterator helper
struct IterT<'a>(&'a dyn T);

// mut iterator helper
struct IterMutT<'a>(&'a mut dyn T);

// iterator implementation
impl<'a> Iterator for IterT<'a> {
    type Item = &'a i32;
    fn next(&mut self) -> Option<Self::Item> {
        // lifetime is inferred just fine
        // self.0 has lifetime 'a
        Some(self.0.get())
    }
}

// mut iterator implementation
impl<'a> Iterator for IterMutT<'a> {
    type Item = &'a mut i32;
    fn next(&mut self) -> Option<Self::Item> {
        // error here: cannot infer an appropriate lifetime
        // self.0 has lifetime of surrounding block, even
        // though it should be 'a as explicitly defined
        Some(self.0.get_mut())
    }
}
```

I expected to see this happen: no error

Instead, this happened: compilation error

```
error[E0495]: cannot infer an appropriate lifetime for autoref due to conflicting requirements
  --> src/lib.rs:30:21
   |
30 |         Some(self.0.get_mut())
   |                     ^^^^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 26:5...
  --> src/lib.rs:26:5
   |
26 | /     fn next(&mut self) -> Option<Self::Item> {
27 | |         // error here: cannot infer an appropriate lifetime
28 | |         // self.0 has lifetime of surrounding block, even
29 | |         // though it should be 'a as explicitly defined
30 | |         Some(self.0.get_mut())
31 | |     }
   | |_____^
note: ...so that reference does not outlive borrowed content
  --> src/lib.rs:30:14
   |
30 |         Some(self.0.get_mut())
   |              ^^^^^^
note: but, the lifetime must be valid for the lifetime `'a` as defined on the impl at 24:6...
  --> src/lib.rs:24:6
   |
24 | impl<'a> Iterator for IterMutT<'a> {
   |      ^^
note: ...so that the types are compatible
  --> src/lib.rs:26:46
   |
26 |       fn next(&mut self) -> Option<Self::Item> {
   |  ______________________________________________^
27 | |         // error here: cannot infer an appropriate lifetime
28 | |         // self.0 has lifetime of surrounding block, even
29 | |         // though it should be 'a as explicitly defined
30 | |         Some(self.0.get_mut())
31 | |     }
   | |_____^
   = note: expected  `std::iter::Iterator`
              found  `std::iter::Iterator`
```

### Meta

I tried it both on stable and nightly rust:

```
$ rustc --version --verbose
rustc 1.45.0-nightly (7ebd87a7a 2020-05-08)
binary: rustc
commit-hash: 7ebd87a7a1e0e21767422e115c9455ef6e6d4bee
commit-date: 2020-05-08
host: x86_64-pc-windows-msvc
release: 1.45.0-nightly
LLVM version: 9.0
```

```
$ rustc --version --verbose
rustc 1.43.0 (4fb7144ed 2020-04-20)
binary: rustc
commit-hash: 4fb7144ed159f94491249e86d5bbd033b5d60550
commit-date: 2020-04-20
host: x86_64-pc-windows-msvc
release: 1.43.0
LLVM version: 9.0
```

**Comment by compiler-errors** [comment]
This is not a bug. You cannot return a mutable longer than the mutable reference you're reborrowing from:

```
error: lifetime may not live long enough
  --> src/lib.rs:30:9
   |
24 | impl<'a> Iterator for IterMutT<'a> {
   |      -- lifetime `'a` defined here
25 |     type Item = &'a mut i32;
26 |     fn next(&mut self) -> Option<Self::Item> {
   |             - let's call the lifetime of this reference `'1`
...
30 |         Some(self.0.get_mut())
   |         ^^^^^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`
```
